File: src/Web/autopilot-monitor-web/app/dashboard/utils/sessionSearchQuery.ts
// Search syntax for the session list's search box.
//
// Same grammar as the event timeline (lib/searchQueryTokens): whitespace-separated
// terms are AND-ed, a leading minus excludes, quotes protect a phrase — plus field
// qualifiers, because "Contoso" and "EliteBook 840" live in two different columns and
// a phrase can never match across a field boundary:
//
//   surface                         substring in any searched field
//   surface failed                  both terms, each in any field
//   model=surface                   only the model field
//   manufacturer="Contoso Ltd"      phrase, only the manufacturer field
//   model="EliteBook 840" -failed   combined with an exclusion
//   "a=b"                           quoted: a literal, not a qualifier
//
// A qualifier that is not in SESSION_SEARCH_FIELDS stays a literal term (`14:30`,
// `foo=bar`). Qualified terms are substring matches like every other term — exact
// matching is the column filter's job. The backend parses the SAME grammar for the
// server-side sweep (`q=` on /api/search/sessions, SessionSearchQuery.cs); both sides
// run utils/session-search-syntax.cases.json, so a term that matches here matches there.

import type { Session } from "../types";
import { literalText, tokenizeSearchQuery } from "@/lib/searchQueryTokens";

export interface SessionSearchField {
  /** The word in front of `=` — lowercase, one per field. */
  readonly qualifier: string;
  readonly key: keyof Session;
  /** Column label, also shown as "Matched: …" in the suggestion list. */
  readonly label: string;
}

/**
 * The searched fields, in suggestion priority order. Exactly the set the backend's
 * free-text predicate searches — a field only one side knows would produce ghost
 * results (server finds it, client filters it back out) or the reverse.
 */
export const SESSION_SEARCH_FIELDS = [
  { qualifier: "device", key: "deviceName", label: "Device" },
  { qualifier: "serial", key: "serialNumber", label: "Serial" },
  { qualifier: "model", key: "model", label: "Model" },
  { qualifier: "manufacturer", key: "manufacturer", label: "Manufacturer" },
  { qualifier: "status", key: "status", label: "Status" },
  { qualifier: "session", key: "sessionId", label: "Session ID" },
  { qualifier: "country", key: "geoCountry", label: "Country" },
  { qualifier: "region", key: "geoRegion", label: "Region" },
  { qualifier: "city", key: "geoCity", label: "City" },
  { qualifier: "agent", key: "agentVersion", label: "Agent Version" },
  { qualifier: "os", key: "osName", label: "OS Name" },
  { qualifier: "build", key: "osBuild", label: "OS Build" },
  { qualifier: "osversion", key: "osDisplayVersion", label: "OS Version" },
  { qualifier: "edition", key: "osEdition", label: "OS Edition" },
  { qualifier: "language", key: "osLanguage", label: "OS Language" },
] as const satisfies readonly SessionSearchField[];

export type SessionSearchQualifier = (typeof SESSION_SEARCH_FIELDS)[number]["qualifier"];
export type SessionSearchKey = (typeof SESSION_SEARCH_FIELDS)[number]["key"];

/** The slice of a session the matcher reads. */
export type SearchableSession = Pick<Session, SessionSearchKey>;

export interface SessionSearchTerm {
  /** Lowercased. */
  text: string;
  /** Set for a qualified term: the one field it may match. */
  field?: SessionSearchKey;
}

export interface ParsedSessionSearchQuery {
  include: SessionSearchTerm[];
  exclude: SessionSearchTerm[];
}

const KEY_BY_QUALIFIER: ReadonlyMap<string, SessionSearchKey> = new Map(
  SESSION_SEARCH_FIELDS.map((f) => [f.qualifier, f.key]),
);

export function parseSessionSearchQuery(query: string): ParsedSessionSearchQuery {
  const include: SessionSearchTerm[] = [];
  const exclude: SessionSearchTerm[] = [];

  for (const token of tokenizeSearchQuery(query)) {
    const field = token.qualifier === undefined ? undefined : KEY_BY_QUALIFIER.get(token.qualifier.toLowerCase());
    const term: SessionSearchTerm = field
      ? { text: token.text.toLowerCase(), field }
      : { text: literalText(token).toLowerCase() };
    const target = token.negated ? exclude : include;
    if (!target.some((t) => t.text === term.text && t.field === term.field)) target.push(term);
  }

  return { include, exclude };
}

function fieldText(session: SearchableSession, key: SessionSearchKey): string {
  return String(session[key] ?? "").toLowerCase();
}

/**
 * Whether one term hits the session. `extra` is client-derived text (formatted date,
 * "N min", "blocked", tenant domain) that only a free term may match — newline-joined by
 * the caller, so a term can never span two of those tokens either.
 */
function termMatches(term: SessionSearchTerm, session: SearchableSession, extra?: string): boolean {
  if (term.field) return fieldText(session, term.field).includes(term.text);
  return (
    SESSION_SEARCH_FIELDS.some((f) => fieldText(session, f.key).includes(term.text)) ||
    (extra !== undefined && extra.toLowerCase().includes(term.text))
  );
}

/**
 * Predicate for the query, or null when it carries no filter (empty, whitespace, a
 * lone minus or a bare `model=` while the user is still typing).
 */
export function buildSessionSearchMatcher(
  query: string,
): ((session: SearchableSession, extra?: string) => boolean) | null {
  const { include, exclude } = parseSessionSearchQuery(query);
  if (include.length === 0 && exclude.length === 0) return null;

  return (session, extra) =>
    include.every((t) => termMatches(t, session, extra)) &&
    !exclude.some((t) => termMatches(t, session, extra));
}

/**
 * For the suggestion list: the first searched field an include term hits, qualified
 * terms first (they name the field the user meant). Null when only `extra` matched
 * (the caller labels that itself) or when the query has no include terms.
 */
export function matchedSearchField(
  parsed: ParsedSessionSearchQuery,
  session: SearchableSession,
): { label: string; value: string } | null {
  const ordered = [...parsed.include.filter((t) => t.field), ...parsed.include.filter((t) => !t.field)];
  for (const term of ordered) {
    for (const f of SESSION_SEARCH_FIELDS) {
      if (term.field && term.field !== f.key) continue;
      const value = session[f.key];
      if (value != null && String(value).toLowerCase().includes(term.text)) {
        return { label: f.label, value: String(value) };
      }
    }
  }
  return null;
}

/**
 * Builds a query of qualified terms in the caller's order — the deep link from Fleet
 * Health's model rows. Empty values are skipped; a value with whitespace or a grammar
 * character is quoted (double quotes inside it are dropped: the grammar has no escape).
 */
export function buildSessionSearchQuery(
  parts: Partial<Record<SessionSearchQualifier, string | null | undefined>>,
): string {
  const terms: string[] = [];
  for (const [qualifier, raw] of Object.entries(parts)) {
    const value = raw?.trim().replace(/"/g, "");
    if (!value || !KEY_BY_QUALIFIER.has(qualifier)) continue;
    terms.push(`${qualifier}=${/[\s=:]/.test(value) ? `"${value}"` : value}`);
  }
  return terms.join(" ");
}

File: src/Backend/AutopilotMonitor.Functions/Helpers/SessionSearchQuery.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using AutopilotMonitor.Shared.Models;

namespace AutopilotMonitor.Functions.Helpers;

/// <summary>
/// The dashboard search-box grammar, parsed once per request for the server-side sweep
/// (<c>q=</c> on <c>/api/search/sessions</c> and the global twin). Mirrors the web's
/// <c>app/dashboard/utils/sessionSearchQuery.ts</c> exactly — both run
/// <c>utils/session-search-syntax.cases.json</c>, because a term that matches on one side
/// and not the other is a ghost result (found by the server, filtered out by the client, or
/// never fetched at all).
/// <para>
/// Grammar: whitespace-separated terms are AND-ed; a leading minus excludes; double quotes
/// protect a phrase; <c>key=value</c> / <c>key:value</c> restricts a term to one field when
/// <c>key</c> is in <see cref="Fields"/> (case-insensitive), otherwise the token stays a
/// literal term (<c>14:30</c>). A qualifier is recognised only when the separator precedes any
/// quote in the token. Every term is a case-insensitive substring match; free terms may hit
/// any field in <see cref="Fields"/>. There is no escape character.
/// </para>
/// </summary>
public sealed class SessionSearchQuery
{
    /// <summary>Qualifier → field reader, in the web's suggestion order. Exactly the set the client searches.</summary>
    public static readonly IReadOnlyList<KeyValuePair<string, Func<SessionSummary, string?>>> Fields =
        new List<KeyValuePair<string, Func<SessionSummary, string?>>>
        {
            new("device", s => s.DeviceName),
            new("serial", s => s.SerialNumber),
            new("model", s => s.Model),
            new("manufacturer", s => s.Manufacturer),
            new("status", s => s.Status.ToString()),
            new("session", s => s.SessionId),
            new("country", s => s.GeoCountry),
            new("region", s => s.GeoRegion),
            new("city", s => s.GeoCity),
            new("agent", s => s.AgentVersion),
            new("os", s => s.OsName),
            new("build", s => s.OsBuild),
            new("osversion", s => s.OsDisplayVersion),
            new("edition", s => s.OsEdition),
            new("language", s => s.OsLanguage),
        };

    private static readonly Dictionary<string, Func<SessionSummary, string?>> ReaderByQualifier =
        Fields.ToDictionary(kv => kv.Key, kv => kv.Value, StringComparer.OrdinalIgnoreCase);

    /// <summary>One search term; <see cref="Field"/> is the qualifier (lowercase) or null for a free term.</summary>
    public readonly record struct Term(string Text, string? Field);

    public IReadOnlyList<Term> Include { get; }
    public IReadOnlyList<Term> Exclude { get; }

    /// <summary>True when the query carries no filter at all (matches every session).</summary>
    public bool IsEmpty => Include.Count == 0 && Exclude.Count == 0;

    private SessionSearchQuery(List<Term> include, List<Term> exclude)
    {
        Include = include;
        Exclude = exclude;
    }

    public static SessionSearchQuery Parse(string? query)
    {
        var include = new List<Term>();
        var exclude = new List<Term>();
        foreach (var token in Tokenize(query ?? string.Empty))
        {
            Term term;
            if (token.Qualifier != null && ReaderByQualifier.ContainsKey(token.Qualifier))
                term = new Term(token.Text.ToString(), token.Qualifier.ToLowerInvariant());
            else
                term = new Term(token.LiteralText, null);
            var target = token.Negated ? exclude : include;
            if (!target.Contains(term)) target.Add(term);
        }
        return new SessionSearchQuery(include, exclude);
    }

    public bool Matches(SessionSummary session)
    {
        if (IsEmpty) return true;
        foreach (var t in Include)
            if (!TermMatches(t, session)) return false;
        foreach (var t in Exclude)
            if (TermMatches(t, session)) return false;
        return true;
    }

    private static bool TermMatches(Term term, SessionSummary session)
    {
        if (term.Field != null)
            return ContainsIgnoreCase(ReaderByQualifier[term.Field](session), term.Text);
        foreach (var field in Fields)
            if (ContainsIgnoreCase(field.Value(session), term.Text)) return true;
        return false;
    }

    private static bool ContainsIgnoreCase(string? haystack, string needle)
        => !string.IsNullOrEmpty(haystack)
           && haystack!.IndexOf(needle, StringComparison.OrdinalIgnoreCase) >= 0;

    private sealed class Token
    {
        public StringBuilder Text { get; } = new();
        public bool Negated { get; init; }
        public string? Qualifier { get; set; }
        public char Separator { get; set; }
        public bool SawQuote { get; set; }
        public string LiteralText => Qualifier == null ? Text.ToString() : $"{Qualifier}{Separator}{Text}";
    }

    // Character-for-character the web tokenizer (lib/searchQueryTokens.ts).
    private static List<Token> Tokenize(string query)
    {
        var tokens = new List<Token>();
        Token? current = null;
        var inQuotes = false;

        void Flush()
        {
            if (current != null && current.Text.Length > 0) tokens.Add(current);
            current = null;
        }

        foreach (var ch in query)
        {
            if (!inQuotes && char.IsWhiteSpace(ch))
            {
                Flush();
                continue;
            }
            if (ch == '"')
            {
                current ??= new Token();
                current.SawQuote = true;
                inQuotes = !inQuotes;
                continue;
            }
            if (current == null)
            {
                if (ch == '-')
                {
                    current = new Token { Negated = true };
                    continue;
                }
                current = new Token();
            }
            if ((ch == '=' || ch == ':') && !inQuotes && !current.SawQuote
                && current.Qualifier == null && current.Text.Length > 0)
            {
                current.Qualifier = current.Text.ToString();
                current.Separator = ch;
                current.Text.Clear();
                continue;
            }
            current.Text.Append(ch);
        }
        Flush();
        return tokens;
    }
}
